src/sqlitedb.cpp

```diff
@@ -196,8 +196,43 @@ bool DBBrowserDB::open(const QString& db, bool readOnly)
     if(tryEncryptionSettings(db, &isEncrypted, &cipherSettings) == false)
         return false;
 
+    // Open database file header to determine encoding
+    QFile file(db);
+    if(!file.open(QIODevice::ReadOnly))
+    {
+        lastErrorMessage = tr("Cannot open database file: %1").arg(file.errorString());
+        return false;
+    }
+
+    QByteArray header = file.read(100);
+    file.close();
+
+    const char* encodingBytes = header.constData() + 56;
+    // The variable 'encoding' will have one of the following values:
+    // 1: UTF-8, 2: UTF-16LE, 3: UTF-16BE
+    quint32 encoding = ((quint8)encodingBytes[0] << 24) |
+                   ((quint8)encodingBytes[1] << 16) |
+                   ((quint8)encodingBytes[2] << 8) |
+                   ((quint8)encodingBytes[3]);
+
     // Open database file
-    if(sqlite3_open_v2(db.toUtf8(), &_db, readOnly ? SQLITE_OPEN_READONLY : SQLITE_OPEN_READWRITE, nullptr) != SQLITE_OK)
+    int status=0;
+    if(encoding == 1)
+    {
+        status = sqlite3_open_v2(
+                db.toUtf8(),
+                &_db,
+                readOnly ? SQLITE_OPEN_READONLY : SQLITE_OPEN_READWRITE,
+                nullptr);
+    }
+    else
+    {
+        status = sqlite3_open16(db.constData(), &_db);
+        if (status == SQLITE_OK && readOnly)
+            sqlite3_exec(_db, "PRAGMA query_only = TRUE;", nullptr, nullptr, nullptr);
```

Comment:
According to https://www.sqlite.org/pragma.html#pragma_query_only
> However, the database is not truly read-only. You can still run a [checkpoint](https://www.sqlite.org/wal.html#ckpt) or a [COMMIT](https://www.sqlite.org/lang_transaction.html) and the return value of the [sqlite3_db_readonly()](https://www.sqlite.org/c3ref/db_readonly.html) routine is not affected.

Is strange that this is the solution for those two issues (#3714 and #3636). Are we working around some bug in SQLite3? Shouldn't it detect the database encoding and act accordingly?

The documentation for these functions say that they determine the default encoding for databases *created* using these functions, not when opening already created databases. Everything seems to indicate that SQLite3 should do that automatically.

There's nothing written, but my impression is that sqlite3_open16 is deprecated in favour of sqlite3_open_v2, and you only have to convert to UTF-8 the filename if running on an OS with UTF-16 filenames (I think Windows is like that, but not Linux; I don't know about macOS). Then, if creating the DB, you use "pragma ENCODING" to set the database encoding.

Reply:
The `sqlite3_open16` function doesn’t provide an option to open the file in read-only mode, unlike `sqlite3_open_v2`, so that’s why it was done this way.

It might be possible to use a URI filename as shown like this: `file:data.db?mode=ro`
but to do so, we would need to compile and provide the SQLite binaries ourselves for Linux, macOS, and Windows.